rasterio/env.py

```diff
@@ -131,46 +132,61 @@ def __init__(self, aws_session=None, aws_access_key_id=None,
+
         self.options = options.copy()
         self.context_options = {}
 
-    def get_aws_credentials(self):
-        """Get credentials and configure GDAL."""
-        import boto3
-        options = {}
-        if not self.aws_session:
-            self.aws_session = boto3.Session(
-                aws_access_key_id=self.aws_access_key_id,
-                aws_secret_access_key=self.aws_secret_access_key,
-                aws_session_token=self.aws_session_token,
-                region_name=self.region_name,
-                profile_name=self.profile_name)
-        self._creds = self.aws_session._session.get_credentials()
-
-        # Pass these credentials to the GDAL environment.
-        if self._creds.access_key:  # pragma: no branch
-            options.update(aws_access_key_id=self._creds.access_key)
-        if self._creds.secret_key:  # pragma: no branch
-            options.update(aws_secret_access_key=self._creds.secret_key)
-        if self._creds.token:
-            options.update(aws_session_token=self._creds.token)
-        if self.aws_session.region_name:
-            options.update(aws_region=self.aws_session.region_name)
-
-        # Pass these credentials to the GDAL environment.
-        local._env.update_config_options(**options)
+        self._creds = None
+
+    @property
+    def is_credentialized(self):
+        return bool(self._creds)
+
+    def credentialize(self):
+        """Get credentials and configure GDAL
+
+        Note well: this method is a no-op if the GDAL environment
+        already has credentials, unless session is not None.
+        """
+        if not hascreds():
+            import boto3
+            if not self.session and not self.aws_access_key_id and not self.profile_name:
+                self.session = boto3.Session()
+            elif not self.session:
+                self.session = boto3.Session(
+                    aws_access_key_id=self.aws_access_key_id,
+                    aws_secret_access_key=self.aws_secret_access_key,
+                    aws_session_token=self.aws_session_token,
+                    region_name=self.region_name,
+                    profile_name=self.profile_name)
+            else:
+                # use self.session
+                pass
+            self._creds = self.session._session.get_credentials()
+
+            # Pass these credentials to the GDAL environment.
+            cred_opts = {}
+            if self._creds.access_key:  # pragma: no branch
+                cred_opts['AWS_ACCESS_KEY_ID'] = self._creds.access_key
+            if self._creds.secret_key:  # pragma: no branch
+                cred_opts['AWS_SECRET_ACCESS_KEY'] = self._creds.secret_key
+            if self._creds.token:
+                cred_opts['AWS_SESSION_TOKEN'] = self._creds.token
+            if self.session.region_name:
+                cred_opts['AWS_REGION'] = self.session.region_name
+            self.options.update(**cred_opts)
+            setenv(**cred_opts)
```

Comment:
A little cleanup. Bug fix is this is a no-op if we've previously credentialized.